nixos: Pulse without pulseaudio

roberth opened the pull request:
###### Description of changes

Make NixOS not care what implements the pulseaudio protocol.

On XFCE this adds the volume control panel widget to systems that run pipewire with pulse enabled.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
nixos/pulseaudio: Add implementation-agnostic `available` option

roberth committed:
nixos/plasma5: Treat pipewire-pulse as pulseaudio

roberth committed:
nixos/xfce: Treat pipewire-pulse as pulseaudio

jansol commented:
:+1: This looks like a step in the right direction. Does it cover all checks for whether pulse is enabled in the config now? Might as well do that tree-wide in this PR.

roberth commented:
I've covered the usages that I was confident about. Still draft, because I have yet to test this on my config though.

jansol commented:
Looking at a quick `rg hardware.pulseaudio.enable` at least the shairport-sync and gdm seem to be checking for whether pulseaudio is enabled but aren't covered by this PR yet.

domenkozar commented:
@roberth do you need help to get this merged?

roberth commented:
@domenkozar I forgot about this PR. I think it just needs a rebase and a review.

jansol commented:
@roberth could you grep for other checks and fix them too? There should be at least the two that I mentioned in my previous comment.

roberth committed:
nixos/shairport-sync: Add pulse group also for pipewire

Thanks to @jansol for pointing out this usage of pulseaudio.enable

Presumably it works, according to https://github.com/mikebrady/shairport-sync/issues/1171

roberth commented:
@jansol

> @roberth could you grep for other checks and fix them too? There should be at least the two that I mentioned in my previous comment.

I've changed shairport, which seems like a safe bet, but I haven't done gdm, because the gdm module seems to do some of the pulse service's work, which may or may not apply to pipewire. I don't use gdm, so I have no idea how to test that.

roberth commented:
I can confirm that it works for xfce.

jansol commented:
> the gdm module seems to do some of the pulse service's work, which may or may not apply to pipewire

Hm, I am not familiar with gdm either.

@hedning @bobby285271 @maxeaubrey you are listed as gdm maintainers, do you happen to have any info on this?

bobby285271 commented:
No idea, maybe someone test [bluetooth audio](https://github.com/NixOS/nixpkgs/pull/61746#issuecomment-493956980) on GNOME? :upside_down_face:

roberth commented on nixos/modules/programs/steam.nix:
```diff
@@ -53,8 +53,8 @@ in {
-    # optionally enable 32bit pulseaudio support if pulseaudio is enabled
-    hardware.pulseaudio.support32Bit = config.hardware.pulseaudio.enable;
+    # optionally enable 32bit pulseaudio support if pulseaudio is available
+    hardware.pulseaudio.support32Bit = config.hardware.pulseaudio.available;
```
Right, so this change is a no-op. I'll revert it.

> set `services.pipewire.alsa.support32Bit` instead.

That will be up to the user. I don't think generalizing this would be particularly helpful, so I don't think this PR can help with that.

JohnRTitor commented on nixos/modules/config/pulseaudio.nix:
```diff
@@ -94,6 +94,19 @@ in {
+        type = types.bool;
+        default = cfg.enable;
+        defaultText = lib.literalMD "Set to `true` by modules that implement PulseAudio when enabled.";
+        description = lib.mdDoc ''
```
```suggestion
        description = ''
```
mdDoc is deprecated

JohnRTitor commented:
Hi @roberth! Are you still working on it? If not, I could take this over, with proper guidance from @jtojnar of course.

roberth commented:
I forgot about this work. My apologies to everyone who provided feedback.

@JohnRTitor I would appreciate that a lot, because I can't make time for this myself. I've been pretty occupied with other things and I've switched to GNOME like a boring person, and I'm not sure if I've hardcoded a workaround or what, but I think refinement of the options like this is still needed to make things work well around pipewire.
I think as a maintainer you can push to my branch, but feel free to use these commits in any way. <!-- including a ritual code burning, or a public roasting, but I was thinking more of cherry-picking -->

JohnRTitor commented:
I have rebased the changes. Anything else I should look for?

emilazy commented:
Is this still relevant after https://github.com/NixOS/nixpkgs/pull/339209?

JohnRTitor commented:
No. Pipewire and its Pulseaudio compatibility layer is default as of https://github.com/NixOS/nixpkgs/pull/339209/commits/438fd16359cf30090c66d3748727829afa90e814

emilazy commented:
But that’s what the PR message implied it helps with:

> On XFCE this adds the volume control panel widget to systems that run pipewire with pulse enabled.

emilazy commented:
Okay, it looks like the only really relevant part of this is the change to `shairport-sync.nix`.

JohnRTitor commented:
Will do it in a separate PR.

JohnRTitor commented:
#341172
